src/main/java/com/sarality/sync/APISyncExecutor.java

```diff
@@ -41,56 +48,78 @@ public void execute() {
     execute(null, null);
   }
 
-  public void execute(TaskProgressListener<SyncProgressCount> progressListener,
+  public boolean execute(TaskProgressListener<SyncProgressCount> progressListener,
       TaskCompletionListener<SyncProgressCount> completionListener) {
     this.progressListener = progressListener;
+    hasErrors = false;
+    syncErrorList = new ArrayList<>();
 
     int pageCount = fetcher.init();
 
-
     List<T> sourceDataList = fetcher.fetchNext();
+    // TODO (@Satya) check if there were errors in fetcher and add to error list.
+
     int currentPage = 0;
     int progressCount = 0;
     int itemsOnPage = 0;
 
     while (sourceDataList != null) {
       List<S> collatedList = collator.collate(sourceDataList);
+      // TODO (@Satya) check if there were errors in collator and add to error list.
       itemsOnPage = collatedList.size();
 
       updateProgress(new SyncProgressCount(++currentPage, pageCount, progressCount, itemsOnPage));
 
       for (S data : collatedList) {
-        R request = requestGenerator.generateSyncRequest(data);
+        R request = null;
+        try {
+          request = requestGenerator.generateSyncRequest(data);
+          // TODO (@Satya) check if there were errors and add them to the error list.
+        } catch (Throwable t) {
+          hasErrors = true;
+          addError(new SyncErrorData(APISyncErrorLocation.API_SYNC_EXECUTE,
+              BaseAPISyncErrorCode.NO_REQUEST_GENERATED, t));
+        }
         if (request != null) {
           try {
             apiExecutor.init(data, request);
           } catch (IOException e) {
-            // TODO (@Satya) if executor could not init - what next?
             logger.info("[SYNC-ERROR] IOError from apiExecutor for request object {} for source {}: " + e.toString(),
                 request.toString(),
                 data.toString());
+            hasErrors = true;
+            addError(new SyncErrorData(APISyncErrorLocation.API_SYNC_EXECUTE,
+                BaseAPISyncErrorCode.IO_EXCEPTION, e));
           }
 
-          if (!apiExecutor.execute()) {
-            // TODO (@Satya) if failed, should increment error count and then determine if we should continue or abort sync
-            // because error count is too high. maintain error/retry queue
-            logger.info("[SYNC-ERROR] Error in API call for request {} for source {}",
-                request.toString(),
-                data.toString());
+          try {
+            if (!apiExecutor.execute()) {
+              // TODO (@Satya) retry based on error type.
+              hasErrors = true;
+              logger.info("[SYNC-ERROR] Error in API call for request {} for source {}",
+                  request.toString(),
+                  data.toString());
+              addErrors(apiExecutor.getSyncErrors());
```

Comment:
I am assuming the same ErrorManagement code is duplicated in the Executor as well

Reply:
Yes, so the ErrorCollector would get used there as well.